### PR TITLE
JIT: change loop inversion edge weight updates and add phase

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Rename `fgOptWhileLoop` to `optInvertWhileLoop` (using terminology from
Muchnick). Split off this transformation into a new phase so it is easier
to see its impact. Make the block updates / reorderings that follow into
a proper phase as well.

Use the test block exit likelihoods to update the profile weights for the
edges involved in loop inversion. Because edge weight updates are now
consistent we no longer need to recompute edge weights afterwards.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

SPMI seems to be missing collections for latest hash. Believe this should be no diff w/o profile data. I see a handful of diffs on a local collection with tiered PGO that I'll post shortly.

**Comment by AndyAyersMS** [comment]
Local diffs (on a collection with tiered PGO enabled)
```
Total bytes of base: 9428
Total bytes of diff: 9217
Total bytes of delta: -211 (-2.24% of base)
    diff is an improvement.


Top file regressions (bytes):
          17 : 74628.dasm (1.66% of base)
          13 : 74645.dasm (0.98% of base)
          11 : 74629.dasm (1.37% of base)
           3 : 74646.dasm (0.34% of base)
           3 : 74651.dasm (0.68% of base)
           3 : 74779.dasm (1.00% of base)
           2 : 4794.dasm (0.68% of base)

Top file improvements (bytes):
        -236 : 74789.dasm (-30.81% of base)
         -12 : 74606.dasm (-1.37% of base)
         -12 : 74580.dasm (-1.41% of base)
          -3 : 4852.dasm (-0.74% of base)

11 total files with Code Size differences (4 improved, 7 regressed), 2 unchanged.

Top method regressions (bytes):
          17 ( 1.66% of base) : 74628.dasm - AssignJagged:first_assignments(System.Int32[][],System.Int16[][]):int
          13 ( 0.98% of base) : 74645.dasm - AssignRect:first_assignments(System.Int32[,],System.Int16[,]):int
          11 ( 1.37% of base) : 74629.dasm - AssignJagged:second_assignments(System.Int32[][],System.Int16[][])
           3 ( 0.34% of base) : 74646.dasm - AssignRect:second_assignments(System.Int32[,],System.Int16[,])
           3 ( 0.68% of base) : 74651.dasm - AssignRect:calc_minimum_costs(System.Int32[,])
           3 ( 1.00% of base) : 74779.dasm - Neural:DoNNetIteration(long):long
           2 ( 0.68% of base) : 4794.dasm - System.Collections.Concurrent.ConcurrentDictionary`2[__Canon,Boolean][System.__Canon,System.Boolean]:TryGetValueInternal(System.__Canon,int,byref):bool:this

Top method improvements (bytes):
        -236 (-30.81% of base) : 74789.dasm - LUDecomp:lubksb(System.Double[][],int,System.Int32[],System.Double[])
         -12 (-1.37% of base) : 74606.dasm - EMFloatClass:DivideInternalFPF(InternalFPF,InternalFPF,InternalFPF)
         -12 (-1.41% of base) : 74580.dasm - EMFloat:DivideInternalFPF(byref,byref,byref)
          -3 (-0.74% of base) : 4852.dasm - Internal.TypeSystem.MethodSignature:Equals(Internal.TypeSystem.MethodSignature,bool):bool:this

Top method regressions (percentages):
          17 ( 1.66% of base) : 74628.dasm - AssignJagged:first_assignments(System.Int32[][],System.Int16[][]):int
          11 ( 1.37% of base) : 74629.dasm - AssignJagged:second_assignments(System.Int32[][],System.Int16[][])
           3 ( 1.00% of base) : 74779.dasm - Neural:DoNNetIteration(long):long
          13 ( 0.98% of base) : 74645.dasm - AssignRect:first_assignments(System.Int32[,],System.Int16[,]):int
           3 ( 0.68% of base) : 74651.dasm - AssignRect:calc_minimum_costs(System.Int32[,])
           2 ( 0.68% of base) : 4794.dasm - System.Collections.Concurrent.ConcurrentDictionary`2[__Canon,Boolean][System.__Canon,System.Boolean]:TryGetValueInternal(System.__Canon,int,byref):bool:this
           3 ( 0.34% of base) : 74646.dasm - AssignRect:second_assignments(System.Int32[,],System.Int16[,])

Top method improvements (percentages):
        -236 (-30.81% of base) : 74789.dasm - LUDecomp:lubksb(System.Double[][],int,System.Int32[],System.Double[])
         -12 (-1.41% of base) : 74580.dasm - EMFloat:DivideInternalFPF(byref,byref,byref)
         -12 (-1.37% of base) : 74606.dasm - EMFloatClass:DivideInternalFPF(InternalFPF,InternalFPF,InternalFPF)
          -3 (-0.74% of base) : 4852.dasm - Internal.TypeSystem.MethodSignature:Equals(Internal.TypeSystem.MethodSignature,bool):bool:this

11 total methods with Code Size differences (4 improved, 7 regressed), 2 unchanged.
```
Suspect the diffs in `lubksb` are not ideal -- downstream of the loop inversion profile data gets compromised, and we lose track of one loop. I plan to fix that in a subsequent change, but can look into fixing it in this PR too.

**Comment by AndyAyersMS** [comment]
Impact of the transformation on `lubksb`. A copy of what's in BB10 was fused into BB01, and the edge weights on BB01 reflect the likelihood of entering the loop or zero-tripping.

```
;; note these are raw counts; charts below show normalized counts

Reducing profile weight of BB10 from 11934.000000 to 11817.000000
Setting weight of BB10 -> BB02 to 11701.146484 (iterate loop)
Setting weight of BB10 -> BB11 to 115.853287 (exit loop)
Profile weight of BB01 remains unchanged at 117.000000
Setting weight of BB01 -> BB02 to 115.852936 (enter loop)
Setting weight of BB01 -> BB11 to 1.147062 (avoid loop)
```

3 other loops get inverted in this example...

<table>
<tr><td>Before</td><td>After</td></tr>
<tr>
<td valign="top">
<td>

</td>
</tr>
</table>

**Comment by AndyAyersMS** [comment]
After the subsequent `optOptimizeLayout` there are some obvious edge weights wrong (eg BB03's outgoing edges):

image - 2021-02-16T132807 350

File: src/coreclr/jit/fgprofile.cpp [hunk 1]
```diff
@@ -3459,4 +3405,145 @@ void Compiler::fgDebugCheckProfileData()
...
+        return false;
+    }
+
+    if (blockWeight > outgoingWeightMax)
+    {
+        JITDUMP("  " FMT_BB " - block weight %f > outgoing max %f\n", block->bbNum, blockWeight, outgoingWeightMax);
+        return false;
+    }
+
+    return missingEdges != 0;
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/fgprofile.cpp:3546]: Good catch... yes the logic is backwards.

**Comment by AndyAyersMS** [comment]
Ran SPMI diffs, found 157 methods with diffs; at a glance they all seem to be from assemblies with IBC.

Am seeing some failures in CI that will require investigation.

File: src/coreclr/jit/compiler.h [hunk 2]
```diff
@@ -6114,7 +6119,8 @@ class Compiler
     void optOptimizeBoolsGcStress(BasicBlock* condBlock);
 #endif
 public:
-    void optOptimizeLayout(); // Optimize the BasicBlock layout of the method
+    PhaseStatus optInvertLoops();    // Invert while loops so they're entered at top and tested at bottom.
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.h:6122]:
```suggestion
    PhaseStatus optInvertLoops();    // Invert loops so they're entered at top and tested at bottom.
```
No need to use `while` here; structure of loops inverted will be described in function comments?

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
`optOptimizeLoops` no longer optimizes anything. I can't resist fixing this. One more push in the works.

**Pull request merged**